File: lecture09/03_minimum/minimum.cpp
// minimum_tree.cpp

#include <iostream>
#include <memory>

using namespace std;

// Node structures

template <typename T>
struct Node {
    T     value;
    Node *left;
    Node *right;
};

// Find minimum of tree

template <typename T>
T minimum_tree(Node<T> *root) {
    T minimum = root->value;

    if (root->left) {
    	minimum = min(minimum, minimum_tree(root->left));
    }

    if (root->right) {
    	minimum = min(minimum, minimum_tree(root->right));
    }

    return minimum;
}

// Main execution

int main(int argc, char *argv[]) {
    Node<int> *root = new Node<int>{7,
    	new Node<int>{5,
    	    new Node<int>{3, nullptr, nullptr},
    	    nullptr
	},
    	new Node<int>{4,
    	    nullptr,
    	    new Node<int>{2, nullptr, nullptr}
	}
    };

    cout << minimum_tree(root) << endl;
    return 0;
}

File: lecture09/03_minimum/minimum.py
#!/usr/bin/env python3

import collections

# Constants

INFINITY = float('inf')

# Structures

Node = collections.namedtuple('Node', 'value left right')

# Functions

def minimum_tree(root):
    ''' Use divide and conquer to compute the minimum of binary tree '''
    # Divide and conquer: recursively solve left and right sub-trees
    left_min  = minimum_tree(root.left)  if root.left  else root.value
    right_min = minimum_tree(root.right) if root.right else root.value

    # Combine: take minimum of current node and left and right minimums
    return min(root.value, left_min, right_min)

# Main Execution

if __name__ == '__main__':
    # Create tree
    root = Node(7,
        Node(5,
                Node(3, None, None),
                None),
        Node(4,
                None,
                Node(2, None, None),
    ))

    # Compute minimum of tree
    print(minimum_tree(root))
